bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/ui/navigator/CSymbolsContentProvider.java

```diff
@@ -13,102 +13,55 @@
-import org.eclipse.lsp4e.outline.SymbolsModel;
+import org.eclipse.core.runtime.jobs.IJobChangeEvent;
+import org.eclipse.core.runtime.jobs.JobChangeAdapter;
+import org.eclipse.jface.viewers.AbstractTreeViewer;
+import org.eclipse.jface.viewers.Viewer;
 import org.eclipse.lsp4e.outline.SymbolsModel.DocumentSymbolWithURI;
-import org.eclipse.lsp4j.DocumentSymbol;
-import org.eclipse.lsp4j.DocumentSymbolParams;
-import org.eclipse.lsp4j.SymbolInformation;
-import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.eclipse.ui.progress.DeferredTreeContentManager;
+import org.eclipse.ui.progress.IDeferredWorkbenchAdapter;
 
 public class CSymbolsContentProvider extends CNavigatorContentProvider {
 
-	class SymbolsContainer {
-		public final IFile file;
-		public final SymbolsModel symbolsModel;
-		public boolean isDirty = true;
+	private final SymbolsManager symbolsManager = SymbolsManager.INSTANCE;
+	private DeferredCSymbolLoader loader;
+	private Object currentInput;
 
-		SymbolsContainer(IFile file) {
-			this.file = file;
-			this.symbolsModel = new SymbolsModel();
-			this.symbolsModel.setUri(file.getLocationURI());
+	@Override
+	public void dispose() {
+		if (currentInput != null && loader != null) {
+			loader.cancel(currentInput);
 		}
+		currentInput = null;
+		symbolsManager.dispose();
+		loader = null;
+		super.dispose();
 	}
 
-	private volatile CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> symbols;
-	private HashMap<URI, SymbolsContainer> cachedSymbols = new HashMap<>();
-
-	private final IFileBufferListener fileBufferListener = new FileBufferListenerAdapter() {
-
-		@Override
-		public void dirtyStateChanged(IFileBuffer buffer, boolean isDirty) {
-			try {
-				// Note: resourceExists must be called prior to buffer.getContentType(),
-				// because getContentType() throws an exception when the underlying resource does not exists.
-				// This can be the case if 'Save as..' has been performed on a file.
-				// Then isDirty is true and dirtyStateChanged gets called for the new, not yet existing file.
-				if (isDirty && resourceExists(buffer) && isCElement(buffer.getContentType())) {
-					var uri = LSPEclipseUtils.toUri(buffer);
-					if (uri != null) {
-						var cachedSymbol = cachedSymbols.get(uri);
-						if (cachedSymbol != null) {
-							cachedSymbol.isDirty = true;
-						}
+	@Override
+	public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
+		if (oldInput != null && loader != null) {
+			loader.cancel(oldInput);
+		}
+		currentInput = newInput;
+		if (viewer instanceof AbstractTreeViewer && newInput != null) {
+			loader = new DeferredCSymbolLoader((AbstractTreeViewer) viewer, (IDeferredWorkbenchAdapter) symbolsManager);
+			loader.addUpdateCompleteListener(new JobChangeAdapter() {
+
+				@Override
+				public void done(IJobChangeEvent event) {
+					if (event.getResult().isOK()) {
+						// Force a selection event
+						viewer.setSelection(viewer.getSelection());
```

Comment:
IIUC by the documentation of IJobChangeListener you cannot rely on it being in the UI thread. The javadoc says `It is undefined in which Thread the notification occurs.` I know other cases in the ecosystem looks like this, but it may no longer be valid since https://github.com/eclipse-platform/eclipse.platform/commit/8186be140279bec9553cffdac31ed851f8344120 and the corresponding N&N entry in https://eclipse.dev/eclipse/news/4.26/platform_isv.php#JobManager_Implementation

Reply:
Thank you for the info. I'll check that.

Reply:
I understand the fix [eclipse-platform/eclipse.platform](https://github.com/eclipse-platform/eclipse.platform/commit/8186be140279bec9553cffdac31ed851f8344120) does not fixed the problem it is undefined in which thread the notification occurs.
 I assume that the problem here is that `viewer.setSelection` or  `viewer.getSelection`  will cause an exception if its not called in a UI thread, right?

Reply:
I removed it